src/extract.rs

```diff
@@ -214,3 +211,508 @@ impl<'a> Extractor<'a> {
+                                .get_mut(&target_sort.name())
+                                .unwrap()
+                                .entry(*target)
+                            {
+                                e.insert((func.decl.name, row.vals.to_vec()));
+                            }
+                        }
+                    }
+                }
+            };
+
+            egraph
+                .backend
+                .dump_table(func.new_backend_id, save_best_parent_edge);
+        }
+    }
+
+    /// This recursively reconstruct the termdag that gives the minimum cost for eclass value.
+    fn reconstruct_termdag_node(
+        &self,
+        egraph: &EGraph,
+        termdag: &mut TermDag,
+        value: &core_relations::Value,
+        sort: &ArcSort,
+    ) -> Term {
+        if sort.is_container_sort() {
+            let elements = sort.inner_values(egraph.backend.containers(), value);
+            let mut ch_terms: Vec<Term> = Vec::new();
+            for ch in elements.iter() {
+                ch_terms.push(self.reconstruct_termdag_node(egraph, termdag, &ch.1, &ch.0));
+            }
+            sort.reconstruct_termdag_container(
+                egraph.backend.containers(),
+                value,
+                termdag,
+                ch_terms,
+            )
+        } else if sort.is_eq_sort() {
+            let (func_name, hyperedge) = self
+                .parent_edge
+                .get(&sort.name())
+                .unwrap()
+                .get(value)
+                .unwrap();
+            let mut ch_terms: Vec<Term> = Vec::new();
+            let ch_sorts = &egraph.functions.get(func_name).unwrap().schema.input;
+            for (value, sort) in hyperedge.iter().zip(ch_sorts.iter()) {
+                ch_terms.push(self.reconstruct_termdag_node(egraph, termdag, value, sort));
+            }
+            termdag.app(*func_name, ch_terms)
+        } else {
+            // Primitive
+            sort.reconstruct_termdag_leaf(egraph.backend.primitives(), value, termdag)
+        }
+    }
+
+    pub fn extract_best_with_sort(
+        &self,
+        egraph: &EGraph,
+        termdag: &mut TermDag,
+        value: core_relations::Value,
+        sort: ArcSort,
+    ) -> Option<(Cost, Term)> {
+        // can be more accurate by considering all sorts extracted but this is ok
+        debug_assert!(self.rootsorts.iter().any(|s| { s.name() == sort.name() }));
+        match self.compute_cost_node(egraph, &value, &sort) {
+            Some(best_cost) => {
+                log::debug!("Best cost for the extract root: {:?}", best_cost);
+
+                let term = self.reconstruct_termdag_node(egraph, termdag, &value, &sort);
+
+                Some((best_cost, term))
+            }
+            None => {
+                log::error!("Unextractable root {:?} with sort {:?}", value, sort,);
+                None
+            }
+        }
+    }
+
+    pub fn extract_best(
```

Comment:
document invariant- extract_best requires a value that matches the rootsort the extractor was created with